core/src/main/scala-3/caliban/schema/ArgBuilderDerivation.scala

```diff
@@ -91,28 +97,69 @@ trait CommonArgBuilderDerivation {
+        case InputValue.ObjectValue(f) if f.size != 1 =>
+          Left(ExecutionError("Exactly one key must be specified for oneOf inputs"))
+      }
+      xs.foldLeft(checkSize)(_ orElse _.partial)
+    }
+
+    def build(input: InputValue): Either[ExecutionError, A] =
+      partial.applyOrElse(input, (in: InputValue) => Left(inputError(in)))
+
+    private def inputError(input: InputValue) =
+      ExecutionError(s"Invalid oneOf input $input for trait $traitLabel")
+  }
+
   private def makeProductArgBuilder[A](
     _fields: => List[(String, ArgBuilder[Any])],
     annotations: Map[String, List[Any]]
-  )(fromProduct: Product => A) = new ArgBuilder[A] {
-    private lazy val fields = _fields
+  )(fromProduct: Product => A): ArgBuilder[A] = new ArgBuilder[A] {
+
+    private val params = Array.from(_fields.map { (label, builder) =>
+      val labelList  = annotations.get(label)
+      val default    = builder.buildMissing(labelList.flatMap(_.collectFirst { case GQLDefault(v) => v }))
+      val finalLabel = labelList.flatMap(_.collectFirst { case GQLName(name) => name }).getOrElse(label)
+      (finalLabel, default, builder)
+    })
+
+    private val required = params.collect { case (label, default, _) if default.isLeft => label }
+
+    override private[schema] val partial: PartialFunction[InputValue, Either[ExecutionError, A]] = {
+      case InputValue.ObjectValue(fields) if required.forall(fields.contains) => fromFields(fields)
+    }
 
     def build(input: InputValue): Either[ExecutionError, A] =
-      fields.view.map { (label, builder) =>
-        input match {
-          case InputValue.ObjectValue(fields) =>
-            val labelList  = annotations.get(label)
-            def default    = labelList.flatMap(_.collectFirst { case GQLDefault(v) => v })
-            val finalLabel = labelList.flatMap(_.collectFirst { case GQLName(name) => name }).getOrElse(label)
-            fields.get(finalLabel).fold(builder.buildMissing(default))(builder.build)
-          case value                          => builder.build(value)
-        }
-      }.foldLeft[Either[ExecutionError, Tuple]](Right(EmptyTuple)) { case (acc, item) =>
-        item match {
-          case Right(value) => acc.map(_ :* value)
-          case Left(e)      => Left(e)
+      input match {
+        case InputValue.ObjectValue(fields) => fromFields(fields)
+        case _                              => Left(ExecutionError("expected an input object"))
+      }
+
+    private def fromFields(fields: Map[String, InputValue]): Either[ExecutionError, A] = {
+      var i          = 0
+      val l          = params.length
+      var acc: Tuple = EmptyTuple
+      while (i < l) {
+        val (label, default, builder) = params(i)
+        val field                     = fields.getOrElse(label, null)
+        val value                     = if (field ne null) builder.build(field) else default
+        value match {
+          case Right(v)    => acc :*= v
```

Comment:
What is this operator?

Reply:
It's shorthand for `acc = acc :* v`, where `:*` is the Scala 3 `Tuple` equivalent of a `Seq`'s `:+` (append)